modules/core/exported/packet.go

```diff
@@ -0,0 +1,34 @@
+package exported
+
+// PacketI defines the standard interface for IBC packets
+type PacketI interface {
+	GetSequence() uint64
+	GetTimeoutHeight() Height
+	GetTimeoutTimestamp() uint64
+	GetSourcePort() string
+	GetSourceChannel() string
+	GetDestPort() string
+	GetDestChannel() string
+	GetData() []byte
+	ValidateBasic() error
+}
+
+// Acknowledgement defines the interface used to return
+// acknowledgements in the OnRecvPacket callback.
+type Acknowledgement interface {
+	Success() bool
+	Acknowledgement() []byte
+}
+
+// AdditionalPacketDataProvider defines the standard interface for retrieving additional packet data.
+// The interface is used to retrieve json encoded data from the packet memo.
+// The interface is also used to retrieve the sender address of the packet.
+type AdditionalPacketDataProvider interface {
+	// GetAdditionalData returns additional packet data keyed by a string.
+	// This function is used to retrieve json encoded data from the packet memo.
+	// If no additional data exists for the key, nil should be returned.
+	GetAdditionalData(key string) map[string]interface{}
+	// GetPacketSender returns the sender address of the packet.
+	// If the packet sender is unknown or undefined, an empty string should be returned.
+	GetPacketSender(srcPortID string) string
```

Comment:
What do you think about making these two separate interfaces. `GetPacketSender` has no expiration date, but `GetAdditionalData` can become obsolete with future IBC changes. I would recommend the following:
```go
// PacketDataProvider defines an optional interfaces for retrieving custom packet data stored on behalf of another application. 
// An existing problem in the IBC middleware design is the inability for a middleware to define its own packet data type and insert packet sender provided information. 
// A short term hack was introduced into several packet data's to utilize a memo to carry this information on behalf of another application. 
// This interfaces standardizes that behaviour. Upon realization of the ability for middleware's to define their own packet data types, this interface will be deprecated and removed with time.  
type PacketDataProvider {
    // GetCustomPacketData returns the packet data held on behalf of another application. 
    // The name the information is stored under should be provided as the key
    // If no custom packet data exists for the key, nil is returned.
    GetCustomPacketData(key string) interface{}
}

// PacketData defines an optional interface which PacketData's may implement.
type PacketData interface {
    // GetPacketSender returns the sender address of the packet. 
    // This should be the address which authorized the packet to be sent. 
    // If the packet sender is unknown or undefined, an empty string should be returned.
    GetPacketSender(sourcePortID string) string
}
```

I chose the name `GetCustomPacketData` because the function is very clearly taking in a key and returning some value. The base packet data here is a carrier for custom packet data information some other application in the stack. I think in the definition of this interface we can clearly state how this is a short term hack to remedy the issue of middleware's being unable to define their own packet data, but state that upon the existence of a flat map which allows other apps to insert their own data, this interface would be removed

Reply:
done